File: sec/aes/aes-cbc-256.js
engine = 'aes-256-cbc'; //aes-256-ecb, aes192, ...
crypto = require('crypto')
log= console.log
ivx = true
class Aes{
    static encryptAes256Cbc(text, _key){
        var cipher;
        let m = crypto.createHash('sha256');
        m.update(_key);
        let key = m.digest();
        let iv = Buffer.from([...Array(16).keys()])
        cipher = crypto.createCipheriv(engine, key, iv);
        //cipher = crypto.createCipher(engine, key); // deprecated: iv 复用风险
        cipher.setAutoPadding(true)
        let ciph = cipher.update(text, 'utf8', 'base64'); //base64, hex ...
        ciph += cipher.final('base64');
        return ciph;
    }

    static decryptAes256Cbc(text, _key){
        var decipher;
        let m = crypto.createHash('sha256');
        m.update(_key);
        let key = m.digest();
        let iv = Buffer.from([...Array(16).keys()])
        decipher = crypto.createDecipheriv(engine, key, iv);
        // decipher.setAutoPadding(false)
        let txt = decipher.update(text,'base64', 'utf8');
        txt  = txt + decipher.final('utf8');
        return txt;
    } 
}

text = '数据'
pass = 'password'
enc = Aes.encryptAes256Cbc(text, pass)
console.log(enc)
dec = Aes.decryptAes256Cbc(enc, pass)
console.log(enc, dec)

File: sec/aes/aes-cbc-256.py
# py-crypt-.md
# pip3 install pycryptodome
import base64
from Crypto.Cipher import AES
from Crypto.Hash import SHA256
from Crypto import Random

def encrypt(key, source):
    source = source.encode()
    key = SHA256.new(key.encode()).digest()  
    IV = bytes(range(16))
    encryptor = AES.new(key, AES.MODE_CBC, IV)
    padding = AES.block_size - len(source) % AES.block_size  
    source += bytes([padding]) * padding 
    data = encryptor.encrypt(source)  
    return base64.b64encode(data).decode("utf8")

def decrypt(key, source):
    source = base64.b64decode(source.encode("utf8"))
    key = SHA256.new(key.encode()).digest()  
    IV = bytes(range(16))
    decryptor = AES.new(key, AES.MODE_CBC, IV)
    data = decryptor.decrypt(source)  
    padding = data[-1]  # pick the padding value from the end;
    if data[-padding:] != bytes([padding]) * padding:
        raise ValueError("Invalid padding...")
    return data[:-padding].decode()  # remove the padding


my_password = "password"
my_data = "数据"

encrypted = encrypt(my_password, my_data)    
decrypted = decrypt(my_password, encrypted)    

print("key:  {}".format(my_password))    
print("data: {}".format(my_data))    
print("enc:  {}".format(encrypted))    
print("dec:  {}".format(decrypted))    
print("data match: {}".format(my_data == decrypted))
